### PR TITLE
WIP fix sideways wiimote

### PR DESCRIPTION
**Pull request by andres-asm**

### CODE REVIEW

**Comment by aliaspider** [comment]
that changes the layout of the normal wiimote too. the descriptors will not match then.
maybe add a conditional for that ?

File: Source/Core/DolphinLibretro/Input.cpp [hunk 1]
```diff
@@ -590,8 +590,10 @@ void retro_set_controller_port_device(unsigned port, unsigned device)
       }
       else
       {
-        wmButtons->SetControlExpression(2, "X");       // 1
-        wmButtons->SetControlExpression(3, "Y");       // 2
+        wmButtons->SetControlExpression(0, "Y");       // 1
+        wmButtons->SetControlExpression(1, "X");       // 2
```
Comment (aliaspider) [inline Source/Core/DolphinLibretro/Input.cpp]: index 0 and 1 would be A and B

**Comment by aliaspider** [comment]
that changes the layout of the normal wiimote too. the descriptors will not match then.
maybe add a conditional for that ?

**Comment by andres-asm** [comment]
I forgot to add WIP.
I'll fix upright when I finish :)

I'm trying to bind tilt now

```diff
diff --git a/Source/Core/DolphinLibretro/Input.cpp b/Source/Core/DolphinLibretro/Input.cpp
index c481be7f69..efc1661555 100644
--- a/Source/Core/DolphinLibretro/Input.cpp
+++ b/Source/Core/DolphinLibretro/Input.cpp
@@ -104,6 +104,8 @@ static struct retro_input_descriptor descWiimote[] = {
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_A, "A"},
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_X, "1"},
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_Y, "2"},
+    {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_L, "Tilt Left" },
+    {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_R, "Tilt Right" },
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_R2, "Shake Wiimote"},
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_START, "+"},
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_SELECT, "-"},
@@ -120,6 +122,8 @@ static struct retro_input_descriptor descWiimoteSideways[] = {
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_A, "2"},
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_X, "A"},
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_Y, "B"},
+    {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_L, "Tilt Left" },
+    {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_R, "Tilt Right" },
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_R2, "Shake Wiimote"},
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_START, "+"},
     {0, RETRO_DEVICE_JOYPAD, 0, RETRO_DEVICE_ID_JOYPAD_SELECT, "-"},
@@ -590,12 +594,16 @@ void retro_set_controller_port_device(unsigned port, unsigned device)
       }
       else
       {
-        wmButtons->SetControlExpression(0, "Y");       // 1
-        wmButtons->SetControlExpression(1, "X");       // 2
+        ControllerEmu::ControlGroup* wmTilt = wm->GetWiimoteGroup(WiimoteGroup::Tilt);
+
+        wmButtons->SetControlExpression(0, "Y");       // B
+        wmButtons->SetControlExpression(1, "X");       // A
         wmButtons->SetControlExpression(2, "B");       // 1
         wmButtons->SetControlExpression(3, "A");       // 2
         wmButtons->SetControlExpression(4, "Select");  // -
         wmButtons->SetControlExpression(5, "Start");   // +
+        wmTilt->SetControlExpression(0, "L");
+        wmTilt->SetControlExpression(1, "R");
       }

       wmButtons->SetControlExpression(6, "R3");                   // Home
```

**Comment by aliaspider** [comment]
~~I think tilt is relative to axis and not direction, so it would be tilt X/Y/Z.~~
Edit: nvm, tilt is defined here : https://github.com/aliaspider/dolphin/blob/master/Source/Core/InputCommon/ControllerEmu/ControlGroup/Tilt.cpp#L22-L31
so left and right would be index 2 and 3.

**Comment by andres-asm** [comment]
ah ok so it would always be 0, (the first argument), then I guess I have to set value somehow?

**Comment by aliaspider** [comment]
the first argument would be 2 for `Tilt Left`  and 3 for `Tilt Right`

**Comment by andres-asm** [comment]
Yeah that works thanks I was looking for that

**Comment by andres-asm** [comment]
Force pushed the changes now I need to test a vertical wiimote game

**Comment by aliaspider** [comment]
maybe I should refactor it into something that looks for the index by name instead.
since this is starting to get confusing xD

**Comment by andres-asm** [comment]
just found an issue... 
this is the API's fault tho (or maybe RetroArch's).

Basically when setting controllers it expects the combined descriptor of all 16 pads.

So if you change the input device for P2 all the P1 descriptors end up as NULL, if you change P3, all P1 and P2 descriptors are NULL and so forth.

This is an issue on every core, what I have done in other cores is generate a final descriptor depending on what all users have selected

So first I generate the possible descriptors
https://github.com/fr500/dosbox-svn/blob/libretro/libretro/libretro_mapper.cpp#L270
And then I keep populating the final descriptor
https://github.com/fr500/dosbox-svn/blob/libretro/libretro/libretro_mapper.cpp#L384

Seems something that should be fixed in the API and the front

**Comment by aliaspider** [comment]
ugh, for real -.- ...
guess that needs some heavy rework then.

**Comment by andres-asm** [comment]
It's not a new problem at least.
Any core would have this issue, say a Genesis core when users have a mix of 3 and 6 button pads, or a PSX core when users have a mix of standard and dualshock pads.

It's just that the most common usercase (SNES) is all users using the default gamepad so noone has bothered to fix this glaring omission :)

**Pull request closed without merging**